src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1,913 +1,1138 @@
+            switch (assignment?.Reference.Type)
+            {
+                case ObjectType objectType:
+                    // lookup declared property
+                    if (objectType.Properties.TryGetValue(propertyName, out var property))
+                    {
+                        return property.TypeReference.Type;
+                    }
+
+                    // if there are additional properties, try those
+                    if (objectType.AdditionalPropertiesType != null)
+                    {
+                        return objectType.AdditionalPropertiesType.Type;
+                    }
+
+                    break;
+
+                case DiscriminatedObjectType discriminated:
+                    if (string.Equals(propertyName, discriminated.DiscriminatorProperty.Name, LanguageConstants.IdentifierComparison))
+                    {
+                        // the property is the discriminator property - use its type
+                        return discriminated.DiscriminatorProperty.TypeReference.Type;
+                    }
+
+                    break;
+            }
+
+            return null;
+        }
+
+        private TypeSymbol? ResolveDiscriminatedObjects(TypeSymbol type, ObjectSyntax syntax)
+        {
+            if (!(type is DiscriminatedObjectType discriminated))
+            {
+                // not a discriminated object type - return as-is
+                return type;
+            }
+
+            var discriminatorProperties = syntax.Properties
+                .Where(p => string.Equals(p.TryGetKeyText(), discriminated.DiscriminatorKey, LanguageConstants.IdentifierComparison))
+                .ToList();
+
+            if (discriminatorProperties.Count != 1)
+            {
+                // the object has duplicate properties with name matching the discriminator key
+                // don't select any of the union members
+                return type;
+            }
+
+            // calling the type check here would prevent the declared type from being assigned to the property
+            // because we haven't yet assigned the declared type to the object
+            // for the purposes of resolving the discriminated object, we just need to check if it's a literal string
+            // which doesn't require the full type check, so we're fine
+            var discriminatorProperty = discriminatorProperties.Single();
+            if (!(discriminatorProperty.Value is StringSyntax stringSyntax))
+            {
+                // the discriminator property value is not a string
+                return type;
+            }
+
+            var discriminatorValue = stringSyntax.TryGetLiteralValue();
+            if (discriminatorValue == null)
+            {
+                // the string value was interpolated
+                return type;
+            }
+
+            // discriminator values are stored in the dictionary as bicep literal string text for some reason

```

Comment:
It's because the key on this dictionary is actually the type name. Should we fix this? I think at least we should probably rename "UnionMembersByKey" to something that more clearly indicates this.

Reply:
Maybe? Ideally, we can have methods that resolve this for the situations that need it. IntelliSense is used much less frequently than the type check and it's currently optimized for the latter, so it's probably ok. I'll update the comment to remove "for some reason", though.